File: 0904-fruit-into-baskets/0904-fruit-into-baskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        int low = 0;
        int res = 0;
        map<int, int> mp;
        for (int high = 0; high < fruits.size(); high++) {
            int right = fruits[high];
            mp[right]++;
            while (mp.size() > 2) {
                int left = fruits[low];
                mp[left]--;
                if (mp[left] == 0) {
                    mp.erase(left);
                }
                low++;
            }
            res = max(res, high - low + 1);
        }
        return res;
    }
};

File: 0904-fruit-into-baskets/0904-fruit-into-baskets.js
/**
 * @param {number[]} fruits
 * @return {number}
 */
var totalFruit = function (fruits) {
    let low = 0;
    let res = 0;
    let map = new Map();
    for (let high = 0; high < fruits.length; high++) {
        let right = fruits[high];
        map.set(right, (map.get(right) || 0) + 1);
        while (map.size > 2) {
            let left = fruits[low];
            map.set(left, map.get(left) - 1);
            if (map.get(left) === 0) map.delete(left);
            low++;
        }
        res = Math.max(res, high-low+1);
    }
    return res;
};